[CS] Move constructor ranking rule into CSRanking

hamishknight opened the pull request:
Previously we were introducing a type variable to mark a constructor's parameter list as `TVO_PrefersSubtypeBinding`. Unfortunately this relies on representing the parameter list as a tuple, which will no longer be properly supported once param flags are removed from tuple types.

Move the logic into CSRanking such that we pick up and compare the parameter lists when comparing overloads. For now, this still relies on comparing the parameter lists as tuples, as there's some subtle tuple subtyping rules that could potentially affect source compatibility here, but at least we can explicitly strip the parameter flags and localise the hack to CSRanking rather than exposing it as a constraint.

This then allows us to remove the `FunctionInput` and `FunctionResult` constraints. The former relied on being able to represent parameter lists as tuples, which won't be possible once parameter flags are stripped from tuple types. The latter is reasonable, but is currently unused.

This should be NFC, I added a couple of constructor ranking test cases in #39481.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

xedin commented on lib/Sema/CSRanking.cpp:
```diff
@@ -766,6 +766,67 @@ static void addKeyPathDynamicMemberOverloads(
+addOverloadBindingTypeDiffs(ConstraintSystem &cs, ConstraintLocator *loc,
+                            ArrayRef<Solution> solutions, unsigned idx1,
+                            unsigned idx2,
+                            SmallVectorImpl<std::pair<Type, Type>> &typeDiff) {
```
Alternatively, could we mark locator for `methodTy` as related to `TypeChecker::getSelfForInitDelegationInConstructor` and extract a an argument tuple from it during "compare type bindings" phase? It seems like that way we don't have to replicate the logic from `matchTupleTypes` here because it would be handled by `isSubtypeOf` call.

hamishknight commented on lib/Sema/CSGen.cpp:
```diff
@@ -1500,25 +1505,12 @@ namespace {
-              ConstraintKind::FunctionInput, methodTy, argTy,
-              CS.getConstraintLocator(expr));
-        }
+        auto memberTypeLoc = CS.getConstraintLocator(
+            expr, LocatorPathElt::ConstructorMemberType(
+                      /*shortFormOrSelfDelegating*/ true));
```
I noticed this locator used to have a `ConstraintLocator::ApplyFunction` element, but given this is on an `UnresolvedDotExpr` anchor that didn't quite seem right to me, so I removed it. Is that correct, or should it be re-added here?

xedin replied:
I think you did the right thing here.

xedin reviewed (approved):
Looks great! Thank you, @hamishknight!

hamishknight commented:
@swift-ci please test source compatibility

xedin commented:
Also I think we can remote it all together once everything is ported to `typeCheckForCodeCompletion`...

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please clean test Linux

hamishknight committed:
[CS] Introduce ConstructorMemberType locator elt

Use this new element to represent the overload type
for a constructor call, and have it store a bit
indicating whether the call is for a short-form
`X(...)` or self-delegating `self.init(...)` call.

hamishknight committed:
[CS] Move constructor ranking rule into CSRanking

Previously we were introducing a type variable
to mark a constructor's parameter list as
`TVO_PrefersSubtypeBinding`. Unfortunately this
relies on representing the parameter list as a
tuple, which will no longer be properly supported
once param flags are removed from tuple types.

Move the logic into CSRanking such that we pick up
and compare the parameter lists when comparing
overload bindings. For now, this still relies on
comparing the parameter lists as tuples, as there's
some subtle tuple subtyping rules that could
potentially affect source compatibility here, but
at least we can explicitly strip the parameter
flags and localise the hack to CSRanking rather
than exposing it as a constraint.

hamishknight committed:
[CS] Fix crasher caught by stress tester

We need to be more lenient checking here as the
type variable may not be bound yet.

hamishknight committed:
[CS] Remove function component constraints

FunctionInput relies on being able to represent
parameter lists as tuples, which won't be possible
once parameter flags are stripped from tuple types.
FunctionResult is reasonable, but is currently
unused.
